bin/redirects/redirect.swift

```diff
@@ -0,0 +1,780 @@
+    (part: "LanguageGuide", filename: "StringsAndCharacters.html", docc: "stringsandcharacters"),
+    (part: "LanguageGuide", filename: "Subscripts.html", docc: "subscripts"),
+    (part: "LanguageGuide", filename: "TheBasics.html", docc: "thebasics"),
+    (part: "LanguageGuide", filename: "TypeCasting.html", docc: "typecasting"),
+
+    (part: "ReferenceManual", filename: "AboutTheLanguageReference.html", docc: "aboutthelanguagereference"),
+    (part: "ReferenceManual", filename: "Attributes.html", docc: "attributes"),
+    (part: "ReferenceManual", filename: "Declarations.html", docc: "declarations"),
+    (part: "ReferenceManual", filename: "Expressions.html", docc: "expressions"),
+    (part: "ReferenceManual", filename: "GenericParametersAndArguments.html", docc: "genericparametersandarguments"),
+    (part: "ReferenceManual", filename: "LexicalStructure.html", docc: "lexicalstructure"),
+    (part: "ReferenceManual", filename: "Patterns.html", docc: "patterns"),
+    (part: "ReferenceManual", filename: "ReferenceManualPart.html", docc: "referencemanual"),
+    (part: "ReferenceManual", filename: "Statements.html", docc: "statements"),
+    (part: "ReferenceManual", filename: "Types.html", docc: "types"),
+    (part: "ReferenceManual", filename: "zzSummaryOfTheGrammar.html", docc: "summaryofthegrammar"),
+
+    (part: "RevisionHistory", filename: "RevisionHistory.html", docc: "revisionhistory"),
+    (part: "RevisionHistory", filename: "RevisionHistoryPart.html", docc: "revisionhistory"),
+]
+
+
+// Values to interpolate into the JavaScript
+var headingRedirects = ""
+var chapterRedirects = ""
+
+for case let (stableID, _, docc) in headings {
+    if docc == "" { continue }
+    headingRedirects += 
+        """
+            "#\(stableID)": "\(docc)",\n
+        """
+}
+
+for case let (part, filename, docc) in chapters {
+    chapterRedirects +=
+        """
+            "\(part)/\(filename)": "\(docc)",\n
+        """
+}
+
+for case let (part, filename, docc) in chapters {
+    let html = 
+    """
+    <html>
+    <head>
+        <title>The Swift Programming Language: Redirect</title>
+        <meta http-equiv="refresh" content="10; url=https://docs.swift.org/swift-book/documentation/the-swift-programming-language/\(docc)">
+    </head>
+    <body>
+    <div>
+    This content has moved; redirecting to the
+    <a href="https://docs.swift.org/swift-book/documentation/the-swift-programming-language/" id="redirect">new location</a>.
+    <noscript>
+    <aside>
+    This page requires JavaScript.
+    Please turn on JavaScript and refresh the page.
+    </aside>
+    </noscript>
+    </div>
+    <script>
+    const headingRedirects = {
+    \(headingRedirects)
+    }
+
+    const chapterRedirects = {
+    \(chapterRedirects)
+    }
+
+    const path = window.location.pathname;
+    const pathComponents = path.split("/");
+    const chapter = pathComponents[pathComponents.length - 1];
```

Comment:
```suggestion
    const chapter = [pathComponents.at(-2), pathComponents.at(-1)].join('/');
```

Because the `chapterRedirects` keys have two parts of a path, I think we need add both parts here or else the later lookup will fail.

Reply:
I think it's safe to look backward two elements here since the root `index.html` doesn't include this script, and all the other pages we used to publish had a section and chapter URL component.

Reply:
I was actually trying to make the opposite change here — because chapter names are unique within the scope of the book, just the last path component is enough — but I need to match that change in `chapterRedirects`.  I don't remember exactly what corner case I found where there might not be two path components at the end — let me see if I can find it again.  Or do you think the `part/chapter` redirect logic as-is would be better?

Reply:
Up to you; I think your original plan sounds better. This was just pointing out that the dictionary lookups are going to fail unless either this variable changes or the dictionary does.

Reply:
Changed the mapping to use filenames only in e95fbb7c.

Reply:
@krilnon Any preference between these spellings it as `pathComponents[pathComponents.length - 1]` versus `pathComponents.at(-1)` ?  I couldn't find the latter when trying to refresh my memory on JavaScript array APIs, which was why I didn't originally use it, but it looks better to me.

Reply:
`pathComponents[pathComponents.length - 1]` is probably compatible with more browsers, [looks like Safari 15.4 was the first revision](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/at) to support `at()`, which is fairly recent.